tests/protocol/test_image_info_matcher.py

```diff
@@ -0,0 +1,105 @@
+        self.assertFalse(test_subject.is_match("Red Hat", "RHEL", "6.3", ""))
+
+    def test_image_exists_by_sku(self):
+        doc = '''{
+            "CANONICAL": {
+                "UBUNTUSERVER": {
+                    "16.04-LTS": { "Match": ".*" }
+                }
+            }
+        }'''
+
+        test_subject = ImageInfoMatcher(doc)
+        self.assertTrue(test_subject.is_match("Canonical", "UbuntuServer", "16.04-LTS", ""))
+        self.assertTrue(test_subject.is_match("Canonical", "UbuntuServer", "16.04-LTS", "16.04.201805090"))
+
+        self.assertFalse(test_subject.is_match("Canonical", "UbuntuServer", "14.04.0-LTS", "16.04.201805090"))
+
+    def test_image_exists_by_version(self):
+        doc = '''{
+            "REDHAT": {
+                "RHEL": {
+                    "Minimum": "6.3"
+                }
+            }
+        }'''
+
+        test_subject = ImageInfoMatcher(doc)
+        self.assertFalse(test_subject.is_match("RedHat", "RHEL", "6.1", ""))
+        self.assertFalse(test_subject.is_match("RedHat", "RHEL", "6.2", ""))
+
+        self.assertTrue(test_subject.is_match("RedHat", "RHEL", "6.3", ""))
+        self.assertTrue(test_subject.is_match("RedHat", "RHEL", "6.4", ""))
+        self.assertTrue(test_subject.is_match("RedHat", "RHEL", "6.5", ""))
+        self.assertTrue(test_subject.is_match("RedHat", "RHEL", "7.0", ""))
+        self.assertTrue(test_subject.is_match("RedHat", "RHEL", "7.1", ""))
+
+    def test_image_exists_by_version01(self):
+        """
+        Test case to ensure the matcher exhaustively searches all cases.
+
+        REDHAT/RHEL have a SKU >= 6.3 is less precise than
+        REDHAT/RHEL/7-LVM have a any version.
+
+        Both should return a successful match.
+        """
+        doc = '''{
+            "REDHAT": {
+                "RHEL": { 
+                    "Minimum": "6.3", 
+                    "7-LVM": { "Match": ".*" }
+                }
+            }
+        }'''
+
+        test_subject = ImageInfoMatcher(doc)
+
+        self.assertTrue(test_subject.is_match("RedHat", "RHEL", "6.3", ""))
+        self.assertTrue(test_subject.is_match("RedHat", "RHEL", "7-LVM", ""))
+
+    def test_ignores_case(self):
+        doc = '''{
+            "CANONICAL": {
+                "UBUNTUSERVER": {
+                    "16.04-LTS": { "Match": ".*" }
+                }
+            }
+        }'''
+
+        test_subject = ImageInfoMatcher(doc)
+        self.assertTrue(test_subject.is_match("canonical", "ubuntuserver", "16.04-lts", ""))
+        self.assertFalse(test_subject.is_match("canonical", "ubuntuserver", "14.04.0-lts", "16.04.201805090"))
+
+    def test_invalid_version(self):
+        doc = '''{
+            "REDHAT": {
+                "RHEL": {
+                    "Minimum": "6.3"
+                }
+            }
+        }'''
+
+        test_subject = ImageInfoMatcher(doc)
+        self.assertFalse(test_subject.is_match("RedHat", "RHEL", "16.04-LTS", ""))
```

Comment:
You also want to test a straightforward numerical comparison, i.e.
```
self.assertFalse(test_subject.is_match("RedHat", "RHEL", "6.04", ""))
```
Or is that supposed to succeed? Darned if I know.

Reply:
I added test cases to account for how FlexibleVersion works. We completely control the specification, so I am not concerned about this type of stuff creeping in.  If it does, there is test coverage to illustrate how it works.